src/librbd/internal.cc

```diff
@@ -1831,14 +1890,28 @@ namespace librbd {
 	Context *req_comp = new C_ContextCompletion(*completion);
 	librados::AioCompletion *rados_completion =
 	  librados::Rados::aio_create_completion(req_comp, NULL, rados_ctx_cb);
+
+	bool flag_nonexistent = false;
 	if (p->offset == 0) {
+	  // TODO: JD if the image as snapshots, do we need to keep the
+	  // object flagged as existing? How does RADOS know to delete
+	  // old snapshot objects?
+	  flag_nonexistent = true;
+	  m_ictx->update_object_map(p->objectno, p->objectno + 1,
+				    OBJECT_PENDING, OBJECT_EXISTS);
 	  m_ictx->data_ctx.aio_remove(p->oid.name, rados_completion);
 	} else {
+	  m_ictx->update_object_map(p->objectno, OBJECT_EXISTS);
 	  librados::ObjectWriteOperation op;
 	  op.truncate(p->offset);
 	  m_ictx->data_ctx.aio_operate(p->oid.name, rados_completion, &op);
 	}
 	rados_completion->release();
+
+	if (flag_nonexistent) {
+	  m_ictx->update_object_map(p->objectno, p->objectno + 1,
```

Comment:
this should be part of the completion passed to aio_remove() so it doesn't finish before the remove (and can use the shorter overload here too)